qualtran/bloqs/mod_arithmetic/mod_subtraction.py

```diff
@@ -0,0 +1,229 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import Dict, Optional, Set, Tuple, TYPE_CHECKING, Union
+
+import sympy
+from attrs import frozen
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BloqBuilder,
+    BloqDocSpec,
+    QBit,
+    QMontgomeryUInt,
+    QUInt,
+    Register,
+    Signature,
+    Soquet,
+    SoquetT,
+)
+from qualtran.bloqs.arithmetic import AddK
+from qualtran.bloqs.basic_gates import XGate
+from qualtran.bloqs.mcmt import And, MultiControlX, MultiTargetCNOT
+from qualtran.drawing import Circle, Text, TextBox, WireSymbol
+from qualtran.symbolics import HasLength
+
+if TYPE_CHECKING:
+    from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+    from qualtran.symbolics import SymbolicInt
+
+
+@frozen
+class ModNeg(Bloq):
+    r"""Performs modular negation.
+
+    Applies the operation $\ket{x} \xrightarrow[]{} \ket{-x \% p}$
+
+    Args:
+        dtype: Datatype of the register.
+        p: The modulus for the negation.
+
+    Registers:
+        x: A dtype input register.
```

Comment:
"the register containing the integer we negate".

Reply:
done